OmegaExamples/src/probmods/generative-models.jl

```diff
@@ -0,0 +1,612 @@
+
+# ╔═╡ cc0c0d70-5197-4b2f-b32d-6ac52b8daf27
+md"`randsample` is used to sample from a random variable:"
+
+# ╔═╡ d607d2e0-84a5-4dfd-a281-f5c64cfe1e2d
+randsample(a)
+
+# ╔═╡ 5a10bc0d-d676-4941-a1d5-41790ef255ec
+md"Run this program a few times. You will get back a different sample on each execution."
+
+# ╔═╡ 02ca238e-5a17-42dc-92a1-be975db0e7cd
+md"To understand how `randsample` works, a key concept to know is the Ω object in Omega.jl."
+
+# ╔═╡ 4ea27371-2e73-4a69-887c-b9f519fbf583
+md"""
+There are a few perspectives on what Ω is:
+
+1. The sample space in probability theory (often denoted $\Omega$)
+2. Space of exogenous variables in as causal graphical modelling nomenclature (often denoted $U$)
+3. The random number generator `rng::Random.AbstractRNG`.  Implementation wise, Omega.jl tries to be compatible with any model written as a function `f(rng::AbstractRNG)`
+"""
+
+# ╔═╡ d70b97b5-af55-46f9-904e-309713603155
+md"""
+Conceptually, sampling from a random variable is done in the following two distinct steps:
+1. Construct a random $ω : Ω$, where $Ω$ can be though of as the space of exogenous variables as in causal graphical modelling nomenclature or simply the random number generator, `rng::Random.AbstractRNG`.
+2. Apply $ω$ to the random variable : $A(ω)$
+"""
+
+# ╔═╡ 0360ccef-ccf4-4284-80a9-ee5ed40becc4
+md"### Primitive and Composite Classes in Omega"
+
+# ╔═╡ 9998d466-eb74-476f-a768-d378bbe83f1b
+md"""##### Primitive Classes
+Omega comes with a set of built-in primitive random variable classes, such as `StdNormal` and `StdUniform`.  There are parameterless infinite sets of random variables.
+"""
+
+# ╔═╡ a0d64a28-e4f9-4003-b69d-7b27bd757c24
+As = StdUniform{Float64}()
+
+# ╔═╡ 84acb184-1c92-4761-a44e-bd1503f8255d
+rv = 1 ~ As
+
+# ╔═╡ 4e80ce71-6757-4908-8cd0-e382b11b70c3
+randsample(rv)
+
+# ╔═╡ 5fd22c50-4c81-41b2-a959-fe3dec00a6fd
+md"""
+##### Composite Class
+
+A class in Omega is actually just function of the form f(id, ω). Of course, you can specify your own classes simply by constructing a function.
+"""
+
+# ╔═╡ 8ac4cad4-eb4c-4e33-815c-e365668b3810
+μ = 2 ~ StdNormal{Float64}()
+
+# ╔═╡ 5fcf6c97-6f81-4be2-b27a-25b2abcabca1
+x(id, ω) = (id ~ Normal(μ(ω), 1))(ω)
+
+# ╔═╡ 1b4d9047-abb0-4fb7-ab72-0a4f790b1515
+x_ = 3 ~ x
+
+# ╔═╡ 0fe48b01-e908-4e6c-8c4c-b68cf8c172dc
+md"`x_` is a random variable of the class `x`"
+
+# ╔═╡ 0a2c3c83-8a67-4666-80c0-f580233dc019
+randsample((μ, x_))
+
+# ╔═╡ f2a20f84-b2e6-4e6c-ab51-6e7ea045954e
+md"Alternatively, `@joint` is used to create the joint distribution of variables:"
+
+# ╔═╡ 7d5587d1-c4b4-48f0-971c-677fbe268e73
+joint = @joint μ x_
+
+# ╔═╡ ae0ad1df-ab2a-4704-a121-623e05f079da
+randsample(joint)
+
+# ╔═╡ 0ae343fa-bc62-48ac-a34e-d21028531530
+md"""
+Every time we create a new (independent) random variable, we need to provide a unique id to it. We could instead use `@~` to generate `id`s automatically. For example, to simulate a (possibly biased) coin toss, we can simply write:
```

Comment:
Again, it would be good to explain the id thing and the ω thing, why they exist and what they do, as early in the doc as possible, or link to a explainer as early as possible, since they are going to be a source of confusion for newcomers to Omega.